state.go

```diff
@@ -317,6 +309,54 @@ func (m *Memberlist) probeNode(node *nodeState) {
 	m.suspectNode(&s)
 }
 
+// Ping initiates a ping to the node with the specified name.
+func (m *Memberlist) Ping(node string, addr net.IP, port uint16) (*time.Duration, error) {
+	ping := ping{SeqNo: m.nextSeqNo(), Node: node}
+
+	// Setup an ack handler.
+	ackCh := make(chan ackMessage, m.config.IndirectChecks+1)
+	m.setAckChannel(ping.SeqNo, ackCh, m.config.ProbeInterval)
+
+	rtt, _, err := m.pingNode(&ping, addr, port, ackCh)
+	return rtt, err
+}
+
+// pingNode handles a single ping with the specified sequence number.
+func (m *Memberlist) pingNode(ping *ping, addr net.IP, port uint16, ackCh chan ackMessage) (*time.Duration, *ackMessage, error) {
+	destAddr := &net.UDPAddr{IP: addr, Port: int(port)}
+
+	// Send the ping message.
+	if err := m.encodeAndSendMsg(destAddr, pingMsg, &ping); err != nil {
+		m.logger.Printf("[ERR] memberlist: Failed to send ping: %s", err)
+		return nil, nil, err
+	}
+
+	// Mark the sent time here, which should be after any pre-processing and
+	// system calls to do the actual send. This probably under-reports a bit,
+	// but it's the best we can do.
+	sent := time.Now()
+
+	// Wait for response or round-trip-time.
+	select {
+	case v := <-ackCh:
+		if v.Complete == true {
+			rtt := v.Timestamp.Sub(sent)
+			return &rtt, &v, nil
+		}
+
+		// As an edge case, if we get a timeout, we need to re-enqueue it
```

Comment:
Hmm - this is referring to a `select()` that's actually up in `probeNode()`. This is kind of tricky given how things are layered here and how the `probeNode()` runs other stuff in parallel with this one. It wouldn't be good to always add it back in at this level.

I'm kind of thinking that given how this code isn't very complex we should just leave `probeNode()` alone and not try to share code. Adding a new `Ping()` method will only be a dozen lines of code, and would keep things simpler overall.